internal/notification/handlers/back_channel_logout.go

```diff
@@ -0,0 +1,278 @@
+	"github.com/zitadel/zitadel/internal/notification/channels/set"
+	_ "github.com/zitadel/zitadel/internal/notification/statik"
+	"github.com/zitadel/zitadel/internal/notification/types"
+	"github.com/zitadel/zitadel/internal/repository/session"
+	"github.com/zitadel/zitadel/internal/repository/sessionlogout"
+	"github.com/zitadel/zitadel/internal/repository/user"
+	"github.com/zitadel/zitadel/internal/user/repository/view"
+	"github.com/zitadel/zitadel/internal/zerrors"
+)
+
+const (
+	BackChannelLogoutNotificationsProjectionTable = "projections.notifications_back_channel_logout"
+)
+
+type backChannelLogoutNotifier struct {
+	commands         *command.Commands
+	queries          *NotificationQueries
+	eventstore       *eventstore.Eventstore
+	authClient       *database.DB
+	keyEncryptionAlg zcrypto.EncryptionAlgorithm
+	channels         types.ChannelChains
+	externalSecure   bool
+	externalPort     uint16
+	idGenerator      id.Generator
+}
+
+func NewBackChannelLogoutNotifier(
+	ctx context.Context,
+	config handler.Config,
+	commands *command.Commands,
+	queries *NotificationQueries,
+	es *eventstore.Eventstore,
+	authClient *database.DB,
+	keyEncryptionAlg zcrypto.EncryptionAlgorithm,
+	channels types.ChannelChains,
+	externalSecure bool,
+	externalPort uint16,
+) *handler.Handler {
+	return handler.NewHandler(ctx, &config, &backChannelLogoutNotifier{
+		commands:         commands,
+		queries:          queries,
+		eventstore:       es,
+		authClient:       authClient,
+		keyEncryptionAlg: keyEncryptionAlg,
+		channels:         channels,
+		externalSecure:   externalSecure,
+		externalPort:     externalPort,
+		idGenerator:      id.SonyFlakeGenerator(),
+	})
+
+}
+
+func (*backChannelLogoutNotifier) Name() string {
+	return BackChannelLogoutNotificationsProjectionTable
+}
+
+func (u *backChannelLogoutNotifier) Reducers() []handler.AggregateReducer {
+	return []handler.AggregateReducer{
+		{
+			Aggregate: session.AggregateType,
+			EventReducers: []handler.EventReducer{
+				{
+					Event:  session.TerminateType,
+					Reduce: u.reduceSessionTerminated,
+				},
+			},
+		}, {
+			Aggregate: user.AggregateType,
+			EventReducers: []handler.EventReducer{
+				{
+					Event:  user.HumanSignedOutType,
+					Reduce: u.reduceUserSignedOut,
+				},
+			},
+		},
+	}
+}
+
+func (u *backChannelLogoutNotifier) reduceUserSignedOut(event eventstore.Event) (*handler.Statement, error) {
+	e, ok := event.(*user.HumanSignedOutEvent)
+	if !ok {
+		return nil, zerrors.ThrowInvalidArgumentf(nil, "HANDL-Gr63h", "reduce.wrong.event.type %s", user.HumanSignedOutType)
+	}
+
+	return handler.NewStatement(event, func(ex handler.Executer, projectionName string) error {
+		ctx, err := u.queries.HandlerContext(event.Aggregate())
+		if err != nil {
+			return err
+		}
+		if !authz.GetFeatures(ctx).EnableBackChannelLogout {
+			return nil
+		}
+		userSession, err := view.UserSessionByIDs(ctx, u.authClient, e.UserAgentID, e.Aggregate().ID, e.Aggregate().InstanceID)
```

Comment:
Wouldn't make it more sense to query the respective session at the point of logout, so at the point of the HumanSIgnedOutEvent creation? So you would not have to query for the session in the processing of the notification.

Reply:
previously there were multiple sessions being queried here but since it's a single session now it makes sense as we anyway need to query on the sign out.
now changed